Tests/NIOPosixTests/IPv4Header.swift

```diff
@@ -0,0 +1,347 @@
+        get {
+            dscpAndEcn >> 2
+        }
+        set {
+            precondition(newValue & 0b0000_0011 == 0)
+            dscpAndEcn = newValue << 2 | (0b0000_0011 & dscpAndEcn)
+            assert(newValue == differentiatedServicesCodePoint)
+        }
+    }
+    var explicitCongestionNotification: UInt8 {
+        get {
+            dscpAndEcn & 0b0000_0011
+        }
+        set {
+            precondition(newValue & 0b0000_0011 == 0)
+            dscpAndEcn = newValue | (0b1111_1100 & dscpAndEcn)
+            assert(newValue == explicitCongestionNotification)
+        }
+    }
+    var totalLength: UInt16
+    var identification: UInt16
+    fileprivate var flagsAndFragmentOffset: UInt16
+    var flags: UInt8 {
+        get {
+            UInt8(flagsAndFragmentOffset >> 13)
+        }
+        set {
+            precondition(newValue & 0b0000_0111 == 0)
+            flagsAndFragmentOffset = UInt16(newValue) << 13 | (0b0001_1111_1111_1111 & flagsAndFragmentOffset)
+            assert(newValue == flags)
+        }
+    }
+    var fragmentOffset: UInt16 {
+        get {
+            flagsAndFragmentOffset & 0b0001_1111_1111_1111
+        }
+        set {
+            precondition(newValue & 0b1110_0000_0000_0000 == 0)
+            flagsAndFragmentOffset = newValue | (0b1110_0000_0000_0000 & flagsAndFragmentOffset)
+            assert(newValue == fragmentOffset)
+        }
+    }
+    var timeToLive: UInt8
+    var `protocol`: NIOIPProtocol
+    var headerChecksum: UInt16
+    var sourceIpAddress: IPv4Address
+    var destinationIpAddress: IPv4Address
+    
+    fileprivate init(
+        versionAndIhl: UInt8,
+        dscpAndEcn: UInt8,
+        totalLength: UInt16,
+        identification: UInt16,
+        flagsAndFragmentOffset: UInt16,
+        timeToLive: UInt8,
+        `protocol`: NIOIPProtocol,
+        headerChecksum: UInt16,
+        sourceIpAddress: IPv4Address,
+        destinationIpAddress: IPv4Address
+    ) {
+        self.versionAndIhl = versionAndIhl
+        self.dscpAndEcn = dscpAndEcn
+        self.totalLength = totalLength
+        self.identification = identification
+        self.flagsAndFragmentOffset = flagsAndFragmentOffset
+        self.timeToLive = timeToLive
+        self.`protocol` = `protocol`
+        self.headerChecksum = headerChecksum
+        self.sourceIpAddress = sourceIpAddress
+        self.destinationIpAddress = destinationIpAddress
+    }
+    
+    init() {
+        self.versionAndIhl = 0
+        self.dscpAndEcn = 0
+        self.totalLength = 0
+        self.identification = 0
+        self.flagsAndFragmentOffset = 0
+        self.timeToLive = 0
+        self.`protocol` = .init(rawValue: 0)
+        self.headerChecksum = 0
+        self.sourceIpAddress = .init(rawValue: 0)
+        self.destinationIpAddress = .init(rawValue: 0)
+    }
+}
+
+
+
+extension ByteBuffer {
+    @discardableResult
+    mutating func readIPv4Header() -> IPv4Header? {
```

Comment:
As discussed offline this should IMHO really not be platform specific.
I think we should have readIPv4Header which reads it correctly. And then we can have another one which is readIPv4HeaderWithDarwinRawSocketQuirks() or something which does the weird thing.
Finally we can then have a readIPv4HeaderFromOSRawSocket()  or something which chooses (in a platform specific way) the right one for the platform

But any readFoo that can't read something written by a writeFoo even if on a different platform is a major red flag

Reply:
There's no reason we shouldn't unit test the darwin quirks on Linux too.

Reply:
Added in https://github.com/apple/swift-nio/pull/2320/commits/1ac14d225b8a8b50f486a1117029d0d313066545